File: frontend/lib/db/opportunities.ts
/**
 * Database queries for opportunities using PostgreSQL directly from Next.js.
 */
import { query } from "./connection";

export interface Opportunity {
  id: string;
  gong_opportunity_id: string;
  name: string;
  account_name: string;
  owner_email: string;
  stage: string;
  close_date: string;
  amount: number;
  health_score: number;
  metadata: any;
  created_at: string;
  updated_at: string;
  call_count?: number;
  email_count?: number;
}

export interface TimelineItem {
  item_type: "call" | "email";
  id: string;
  gong_call_id?: string;
  gong_email_id?: string;
  timestamp: string;
  title?: string;
  subject?: string;
  duration?: number;
  sender_email?: string;
}

export async function searchOpportunities({
  filters = {},
  sort = "updated_at",
  sortDir = "DESC",
  limit = 50,
  offset = 0,
}: {
  filters?: Record<string, any>;
  sort?: string;
  sortDir?: string;
  limit?: number;
  offset?: number;
}): Promise<{ opportunities: Opportunity[]; total: number }> {
  // Build WHERE clauses
  const whereClauses: string[] = [];
  const params: any[] = [];
  let paramIndex = 1;

  if (filters.owner) {
    whereClauses.push(`o.owner_email = $${paramIndex++}`);
    params.push(filters.owner);
  }

  if (filters.stage) {
    if (Array.isArray(filters.stage)) {
      whereClauses.push(`o.stage = ANY($${paramIndex++})`);
      params.push(filters.stage);
    } else {
      whereClauses.push(`o.stage = $${paramIndex++}`);
      params.push(filters.stage);
    }
  }

  if (filters.health_score_min !== undefined) {
    whereClauses.push(`o.health_score >= $${paramIndex++}`);
    params.push(filters.health_score_min);
  }

  if (filters.health_score_max !== undefined) {
    whereClauses.push(`o.health_score <= $${paramIndex++}`);
    params.push(filters.health_score_max);
  }

  if (filters.search) {
    whereClauses.push(
      `(o.name ILIKE $${paramIndex} OR o.account_name ILIKE $${paramIndex})`
    );
    params.push(`%${filters.search}%`);
    paramIndex++;
  }

  const whereSQL = whereClauses.length > 0 ? `WHERE ${whereClauses.join(" AND ")}` : "";

  // Validate sort field
  const validSorts = new Set(["updated_at", "close_date", "health_score", "amount"]);
  const sortField = validSorts.has(sort) ? sort : "updated_at";
  const sortDirection = sortDir.toUpperCase() === "ASC" ? "ASC" : "DESC";

  // Get total count
  const countResult = await query(
    `SELECT COUNT(*) as total FROM opportunities o ${whereSQL}`,
    params
  );
  const total = parseInt(countResult.rows[0]?.total || "0", 10);

  // Get paginated results
  params.push(limit, offset);
  const result = await query(
    `
    SELECT
      o.*,
      COUNT(DISTINCT co.call_id) as call_count,
      COUNT(DISTINCT e.id) as email_count
    FROM opportunities o
    LEFT JOIN call_opportunities co ON o.id = co.opportunity_id
    LEFT JOIN emails e ON o.id = e.opportunity_id
    ${whereSQL}
    GROUP BY o.id
    ORDER BY o.${sortField} ${sortDirection} NULLS LAST
    LIMIT $${paramIndex} OFFSET $${paramIndex + 1}
    `,
    params
  );

  return {
    opportunities: result.rows,
    total,
  };
}

export async function getOpportunity(id: string): Promise<Opportunity | null> {
  const result = await query(
    `
    SELECT
      o.*,
      COUNT(DISTINCT co.call_id) as call_count,
      COUNT(DISTINCT e.id) as email_count
    FROM opportunities o
    LEFT JOIN call_opportunities co ON o.id = co.opportunity_id
    LEFT JOIN emails e ON o.id = e.opportunity_id
    WHERE o.id = $1
    GROUP BY o.id
    `,
    [id]
  );

  return result.rows[0] || null;
}

export async function getOpportunityTimeline({
  opportunityId,
  limit = 20,
  offset = 0,
}: {
  opportunityId: string;
  limit?: number;
  offset?: number;
}): Promise<{ items: TimelineItem[]; total: number }> {
  // Get total count
  const countResult = await query(
    `
    SELECT COUNT(*) as total FROM (
      SELECT c.id FROM calls c
      JOIN call_opportunities co ON c.id = co.call_id
      WHERE co.opportunity_id = $1
      UNION ALL
      SELECT e.id FROM emails e
      WHERE e.opportunity_id = $1
    ) timeline
    `,
    [opportunityId]
  );
  const total = parseInt(countResult.rows[0]?.total || "0", 10);

  // Get timeline items
  const result = await query(
    `
    SELECT
      'call' as item_type,
      c.id,
      c.gong_call_id,
      c.title,
      c.scheduled_at as timestamp,
      c.duration,
      NULL as subject,
      NULL as sender_email,
      NULL as gong_email_id
    FROM calls c
    JOIN call_opportunities co ON c.id = co.call_id
    WHERE co.opportunity_id = $1

    UNION ALL

    SELECT
      'email' as item_type,
      e.id,
      NULL as gong_call_id,
      NULL as title,
      e.sent_at as timestamp,
      NULL as duration,
      e.subject,
      e.sender_email,
      e.gong_email_id
    FROM emails e
    WHERE e.opportunity_id = $1

    ORDER BY timestamp DESC
    LIMIT $2 OFFSET $3
    `,
    [opportunityId, limit, offset]
  );

  return {
    items: result.rows,
    total,
  };
}

File: api/v1/opportunities.py
"""
Opportunity API Endpoints

Provides REST API access to opportunity data including:
- Listing opportunities with filters and pagination
- Fetching opportunity details
- Accessing opportunity timeline (calls and emails)
"""

import logging
from typing import Any, Literal

from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel, Field

from db import queries

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/opportunities", tags=["opportunities"])


# ============================================================================
# REQUEST/RESPONSE MODELS
# ============================================================================


class OpportunityListResponse(BaseModel):
    """Response model for opportunity list endpoint."""

    opportunities: list[dict[str, Any]] = Field(description="List of opportunities")
    total: int = Field(description="Total count of opportunities matching filters")
    page: int = Field(description="Current page number (1-indexed)")
    page_size: int = Field(description="Number of items per page")
    has_more: bool = Field(description="Whether there are more pages available")


class OpportunityDetailResponse(BaseModel):
    """Response model for opportunity detail endpoint."""

    opportunity: dict[str, Any] = Field(description="Opportunity details")
    timeline: list[dict[str, Any]] = Field(description="Timeline of calls and emails", default=[])


# ============================================================================
# ENDPOINTS
# ============================================================================


@router.get("", response_model=OpportunityListResponse)
async def list_opportunities(
    owner_email: str | None = Query(None, description="Filter by opportunity owner email"),
    stage: str | None = Query(None, description="Filter by stage (e.g., 'Prospecting')"),
    health_score_min: float | None = Query(
        None, description="Minimum health score (0.0-10.0)", ge=0.0, le=10.0
    ),
    health_score_max: float | None = Query(
        None, description="Maximum health score (0.0-10.0)", ge=0.0, le=10.0
    ),
    search: str | None = Query(None, description="Search in opportunity name and account name"),
    sort: Literal["updated_desc", "close_date_asc", "amount_desc", "name_asc"] = Query(
        "updated_desc", description="Sort order"
    ),
    page: int = Query(1, description="Page number (1-indexed)", ge=1),
    page_size: int = Query(20, description="Items per page", ge=1, le=100),
) -> OpportunityListResponse:
    """
    List opportunities with optional filters and pagination.

    Supports filtering by:
    - Owner email
    - Stage (e.g., Prospecting, Qualification, Proposal, Negotiation, Closed Won)
    - Health score range
    - Search text (name, account)

    Supports sorting by:
    - updated_desc: Most recently updated first (default)
    - close_date_asc: Soonest closing date first
    - amount_desc: Highest value first
    - name_asc: Alphabetical by name

    Returns paginated results with total count and pagination metadata.
    """
    try:
        # Build filters dict
        filters: dict[str, Any] = {}
        if owner_email:
            filters["owner_email"] = owner_email
        if stage:
            filters["stage"] = stage
        if health_score_min is not None:
            filters["health_score_min"] = health_score_min
        if health_score_max is not None:
            filters["health_score_max"] = health_score_max
        if search:
            filters["search"] = search

        # Calculate offset from page number
        offset = (page - 1) * page_size

        # Query database
        result = queries.search_opportunities(
            filters=filters, sort=sort, limit=page_size, offset=offset
        )

        # Calculate pagination metadata
        total = result["total"]  # type: ignore[call-overload,index]
        has_more = offset + page_size < total

        return OpportunityListResponse(
            opportunities=result["opportunities"],  # type: ignore[call-overload,index]
            total=total,
            page=page,
            page_size=page_size,
            has_more=has_more,
        )

    except Exception as e:
        logger.error(f"Error listing opportunities: {e}", exc_info=True)
        raise HTTPException(
            status_code=500,
            detail=f"Failed to list opportunities: {str(e)}",
        ) from e


@router.get("/{opportunity_id}", response_model=OpportunityDetailResponse)
async def get_opportunity_detail(
    opportunity_id: str,
    timeline_limit: int = Query(50, description="Number of timeline items to return", ge=1, le=200),
    timeline_offset: int = Query(0, description="Timeline pagination offset", ge=0),
) -> OpportunityDetailResponse:
    """
    Get detailed information about a specific opportunity.

    Returns:
    - Opportunity metadata (name, account, owner, stage, close date, amount, health score)
    - Counts of associated calls and emails
    - Timeline of calls and emails (paginated)

    Timeline items are sorted chronologically (newest first) and include:
    - Calls: title, date, participants, coaching scores
    - Emails: subject, sender, recipients, sent date, body snippet
    """
    try:
        # Get opportunity details
        opportunity = queries.get_opportunity(opportunity_id)

        if not opportunity:
            raise HTTPException(
                status_code=404,
                detail=f"Opportunity not found: {opportunity_id}",
            )

        # Get timeline items
        timeline_result = queries.get_opportunity_timeline(
            opp_id=opportunity_id, limit=timeline_limit, offset=timeline_offset
        )

        return OpportunityDetailResponse(
            opportunity=opportunity, timeline=timeline_result["items"]  # type: ignore[call-overload,index]
        )

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error fetching opportunity {opportunity_id}: {e}", exc_info=True)
        raise HTTPException(
            status_code=500,
            detail=f"Failed to fetch opportunity: {str(e)}",
        ) from e
